File: src/main/java/net/guneyilmaz0/permissionss/listener/PlayerListener.java
package net.guneyilmaz0.permissionss.listener;

import cn.nukkit.Player;
import cn.nukkit.Server;
import cn.nukkit.event.EventHandler;
import cn.nukkit.event.Listener;
import cn.nukkit.event.player.PlayerChatEvent;
import cn.nukkit.event.player.PlayerJoinEvent;
import cn.nukkit.event.player.PlayerQuitEvent;
import cn.nukkit.permission.PermissionAttachment;
import net.guneyilmaz0.permissionss.Utils;
import net.guneyilmaz0.permissionss.event.PlayerGroupChangeEvent;
import net.guneyilmaz0.permissionss.event.PlayerRankExpiredEvent;
import net.guneyilmaz0.permissionss.group.Group;
import net.guneyilmaz0.permissionss.group.GroupManager;
import net.guneyilmaz0.permissionss.permission.PermissionManager;

@SuppressWarnings("unused")
public class PlayerListener implements Listener {

    @EventHandler
    public void onJoin(PlayerJoinEvent event) {
        Player player = event.getPlayer();
        PermissionManager.handlePermissions(player);
        player.setNameTag(GroupManager.getPlayerGroup(player).getNameTagFormat().replaceAll("%nickname%", player.getName()));
    }

    @EventHandler
    public void onQuit(PlayerQuitEvent event) {
        PermissionAttachment attachment = PermissionManager.permissions.get(event.getPlayer().getUniqueId());
        if (attachment == null) return;
        try {
            event.getPlayer().removeAttachment(attachment);
        } catch (Throwable ignored) {}
    }


    @EventHandler
    public void onChat(PlayerChatEvent event) {
        Player player = event.getPlayer();
        Group group = GroupManager.getPlayerGroup(player);
        String chatFormat = group.getChatFormat();
        String message = chatFormat.replaceAll("%nickname%", player.getName()).replaceAll("%message%", event.getMessage());
        event.setFormat(message);
    }

    @EventHandler
    public void onGroupChange(PlayerGroupChangeEvent event) {
        Player player = event.getPlayer();
        player.setNameTag(event.getNewGroup().getNameTagFormat());
        player.sendMessage(Utils.translate("commands.set_group.player_message", event.getNewGroup().getName()));
    }

    @EventHandler
    public void onRankExpired(PlayerRankExpiredEvent event) {
        Player player = event.getPlayer();
        Group oldGroup = GroupManager.getPlayerGroup(player);
        Group group = GroupManager.getDefaultGroup();
        GroupManager.setPlayerGroup(player, group);
        PlayerGroupChangeEvent groupChangeEvent = new PlayerGroupChangeEvent(player, group, oldGroup, null);
        Server.getInstance().getPluginManager().callEvent(event);
    }
}

File: src/main/kotlin/listeners/PlayerListener.kt
package net.guneyilmaz0.permissions.listeners

import cn.nukkit.Server
import cn.nukkit.event.EventHandler
import cn.nukkit.event.Listener
import cn.nukkit.event.player.PlayerChatEvent
import cn.nukkit.event.player.PlayerJoinEvent
import cn.nukkit.event.player.PlayerQuitEvent
import net.guneyilmaz0.permissions.Utils
import net.guneyilmaz0.permissions.events.PlayerGroupChangeEvent
import net.guneyilmaz0.permissions.events.PlayerRankExpiredEvent
import net.guneyilmaz0.permissions.group.GroupManager
import net.guneyilmaz0.permissions.permission.PermissionManager

@Suppress("unused")
class PlayerListener : Listener {

    @EventHandler
    private fun onJoin(event: PlayerJoinEvent) {
        val player = event.player
        PermissionManager.handlePermissions(player)
        player.nameTag = GroupManager.getPlayerGroup(player)?.nameTagFormat?.replace("%nickname%", player.name) ?: player.name
    }

    @EventHandler
    private fun onQuit(event: PlayerQuitEvent) {
        val attachment = PermissionManager.permissions[event.player.uniqueId] ?: return
        event.player.removeAttachment(attachment)
    }

    @EventHandler
    private fun onChat(event: PlayerChatEvent) {
        val player = event.player
        val group = GroupManager.getPlayerGroup(player) ?: return
        val format = group.chatFormat
        val message = format.replace("%nickname%", player.name).replace("%message%", event.message)
        event.format = message
    }

    @EventHandler
    private fun onGroupChange(event: PlayerGroupChangeEvent) {
        val player = event.player
        player.nameTag = event.newGroup.nameTagFormat.replace("%nickname%", player.name)
        player.sendMessage(Utils.translate("commands.set_group.player_message", event.newGroup.name))
    }

    @EventHandler
    private fun onRankExpired(event: PlayerRankExpiredEvent) {
        val player = event.player
        val group = GroupManager.getDefaultGroup()
        GroupManager.setPlayerGroup(player, group)
        val groupChangeEvent = PlayerGroupChangeEvent(player, group, player)
        Server.getInstance().pluginManager.callEvent(groupChangeEvent)
    }
}